src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -137,53 +137,59 @@ SymbolKind.NetModule or
                     if (state is null)
                         return null;
 
-                    // Determine the old/stale compilation to help seed the non-final state with.
-                    var staleCompilationWithGeneratedDocuments = state switch
+                    if (state is FinalCompilationTrackerState finalState)
                     {
-                        FinalCompilationTrackerState finalState => finalState.FinalCompilationWithGeneratedDocuments,
-                        NonFinalWithCompilationTrackerState nonFinalState => nonFinalState.StaleCompilationWithGeneratedDocuments,
-                        _ => throw ExceptionUtilities.UnexpectedValue(state.GetType()),
-                    };
-
-                    var intermediateProjects = UpdateIntermediateProjects(oldProjectState, state, translate);
-
-                    var newState = CompilationTrackerState.Create(
-                        state.IsFrozen,
-                        state.CompilationWithoutGeneratedDocuments,
-                        state.GeneratorInfo,
-                        staleCompilationWithGeneratedDocuments,
-                        intermediateProjects);
-                    return newState;
+                        // Roll back from the final state to an in progress state.
+                        var newState = InProgressState.Create(
+                            finalState.IsFrozen,
+                            finalState.CompilationWithoutGeneratedDocuments,
+                            finalState.GeneratorInfo,
+                            finalState.FinalCompilationWithGeneratedDocuments,
```

Comment:
Nit: consider named args for the compilation inputs just to ensure we don't cross them.

Reply:
done.